storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -1024,22 +1074,50 @@ private Optional<SlotAndExecutionPayloadSummary> updateFinalizedOptimisticTransi
   private void removeNonCanonicalBlobSidecars(
       final Map<Bytes32, UInt64> deletedHotBlocks,
       final Map<Bytes32, Bytes32> finalizedChildToParentMap) {
-    try (final FinalizedUpdater updater = finalizedUpdater()) {
+
+    final Set<SlotAndBlockRoot> nonCanonicalBlocks =
+        deletedHotBlocks.entrySet().stream()
+            .filter(entry -> !finalizedChildToParentMap.containsKey(entry.getKey()))
+            .map(entry -> new SlotAndBlockRoot(entry.getValue(), entry.getKey()))
+            .collect(Collectors.toSet());
+
+    if (storeNonCanonicalBlocks) {
+      final Iterator<SlotAndBlockRoot> nonCanonicalBlocksIterator = nonCanonicalBlocks.iterator();
+      int index = 0;
+      while (nonCanonicalBlocksIterator.hasNext()) {
+        int start = index;
+        try (final FinalizedUpdater updater = finalizedUpdater()) {
+          while (nonCanonicalBlocksIterator.hasNext() && (index - start) < BLOBS_TX_BATCH_SIZE) {
+            dao.getBlobSidecarKeys(nonCanonicalBlocksIterator.next())
+                .forEach(
+                    key -> {
+                      dao.getBlobSidecar(key)
+                          .ifPresent(
+                              blobSidecarBytes -> {
+                                BlobSidecar blobSidecar =
+                                    spec.deserializeBlobSidecar(blobSidecarBytes, key.getSlot());
+                                updater.addNonCanonicalBlobSidecar(blobSidecar);
```

Comment:
we could avoid this deserialization if we add a updater.addNonCanonicalBlobSidecarRaw that uses `putRaw` under the hood. So we can tranfer bytes only and avoid CPU pressure for a deserialization-serialization cycle.

Reply:
Definitely makes sense. I added a `addNonCanonicalBlobSidecarRaw` method for that